context.go

```diff
@@ -337,6 +361,39 @@ func (c *context) Verify(resource Resource) error {
 	return nil
 }
 
+func (c *context) checkoutFile(fp string, rf RegularFile, flag int) error {
+	if c.provider == nil {
+		return fmt.Errorf("no file provider")
+	}
+	var (
+		r   io.Reader
+		err error
+	)
+	for _, dgst := range rf.Digests() {
+		r, err = c.provider.Reader(dgst)
+		if err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return fmt.Errorf("file content could not be provided: %v", err)
+	}
+
+	f, err := os.OpenFile(fp, flag, rf.Mode())
```

Comment:
Needs something like this: https://github.com/docker/swarmkit/blob/f0983eb859019cf5dcfef9ef3baa2facb8a1c4f5/ioutils/ioutils.go#L13.

Basically, make a _sibling_ tempfile, write all data, flush, then move.